### ISSUE TITLE
Balanced Grain Policy: add lookback parameter

### ISSUE THREAD

**Issue by blueridger** [sourcecred/sourcecred#2678]
Similar to https://github.com/sourcecred/sourcecred/pull/2660/files it would be nice to have a lookback parameter on the balanced policy.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200032326400385/1200038311637562) by [Unito](https://www.unito.io/learn-more)

**Comment by elihanover** [comment]
I could use a "window" with `start` and `end` intervals, where this lookback version is the instance where end = the most recent.  This would let us apply this Balanced policy to whatever interval we'd like.  Thoughts @blueridger ?

Also, more generally we might want this window for _all_ policies, which makes me think it would be cleaner to implement this window feature as something that can be applied to any policy (as opposed to rewriting the logic in each one).  This isn't in scope for this, but I've noticed that modifying and creating new policies is clunky, and I'd argue it's a high leverage feature to allow communities to create distribution policies easily.  Not pressing, but I'd be interested in cracking this if there's mutual interest from others.

**Comment by blueridger** [comment]
That strategy for the intervals sounds good. Defining an interface for 3rd party policies is a cool idea. I really want to pluginize grain more and more. I imagine our plugin contract refactor that we're about to do will inform such a project. For now, we can actually pretty easily combine the functionalities of our 3 policies into 1 which is better than investing further in the separate policies. I'd focus on that as a high-impact incremental improvement, personally. And I'd wait until after the plugin contract refactor and a grain distribution interface get defined before investing in refactoring the look back to be generically applyable.

**Comment by bensodenkamp** [comment]
So a couple questions about expected behavior:

1. @elihanover  Regarding defining a window with a start and end date: this seems problematic because if the end date is before the last distribution, the present distribution will not take into account the last distribution. In other words, if there is a delay between when we try to balance grain awards, and when we include information about the most recent grain awards, we'll be stuck trying to balance something we already tried to balance at least once without being able to see the results of our last attempt.

2. What is the use case we are trying to address with this feature? are we trying to:
  2.1. Balance all time grain awards against limited time cred?
  2.2. Balance limited time grain awards against limited time cred?

Doing 2.2 is a lot more work because the plumbing is not currently there to evaluate distributions of grain over a limited time window, only awards of cred. If we wanted to evaluate grain over a limited time window, we would have to pull that data out of the ledger when we went to evaluated the policies.

On one hand, it seems wrong to compare all time grain vs limited time cred, but on the other hand grain is always delayed from cred, so even if you are computing limited time grain, you don't really know what cred/config defined the state when that grain was awarded.

In short, what exact problem are we solving?

**Comment by bensodenkamp** [comment]
Thinking about the use case further, it seems like limiting the lookback for the balanced policy would have the effect of: 

If something you did a bit ago turns out the be more valuable in recent days, you'll get more grain to balance the additional retroactive cred, but if you did the valuable thing a long time ago, it won't give you any more grain even if it gave you more cred.

Is this the intention? If so it seems like we need to limit the time window for grain as well as cred otherwise you could get:

Seems like you got paid a lot of grain some time in the past but the system can no longer see the cred that resulted in that grain so it's sure you're super overpaid, and should get any grain for the things you're doing now.

...which would be a problem.

**Comment by blueridger** [comment]
Excellent questions. We've been wrestling that in our reporting features too. We don't have a good answer yet on how to conceptualize localized grain-per-cred. I think the point of this issue is that historical cred is kinda bull, because we don't have time-scoped weights so when weight changes happen and no longer align with the historical crediquette then historical cred is super messed up. But the goal is `2.2. Balance limited time grain awards against limited time cred?`

One thing to consider is that distributions have a "credTimestamp" attribute which is the cred interval associated and is different from the distribution event's timestamp. So we could consider solutions where the balanced policy is actually creating contemporary distributions across historical intervals. For example, if someone is relatively underpaid looking at aggregrate grain and cred distributed for the last 3 weeks, and it's primarily due to being underpaid 3 weeks ago, the policy could create a new distribution with a credTimestamp for 3 weeks ago.

The plumbing would need refactored as you say, ideally piping the CredGrainView all the way down rather than converting it to CredAccounts. This is a pretty big increase in scope, but it's not monstrous.

**Comment by blueridger** [comment]
However this line would need changed because it assumes distributions appear in the ledger in ascending credTimestamp order.

https://github.com/sourcecred/sourcecred/blob/main/packages/sourcecred/src/core/ledger/ledger.js#L983

**Comment by bensodenkamp** [comment]
OK cool so with this strategy, the balanced policy would be changed to

1. Step through each historical interval in the look back (or forever if a lookback is not defined)
2. For each unique cred interval, evaluate the difference between the present calculation of cred at that interval and the grain awarded based on that interval (as associated by the credTimestamp)
3. Apply the budget for balanced policy proportionally across identities on an oldest-imbalance-first basis (because those would be the first imbalances to expire as the window slides) by creating contemporary distributions which are date-associated with the intervals they attempt to balance.
  3.1 This would involve iterating through the intervals oldest to newest and applying as much of the budget is still available to imbalances for each interval until the budget is gone.

Sound right?

**Comment by bensodenkamp** [comment]
I see the work of getting grain for a cred interval is already done:

static _calculateGrainEarnedPerInterval(
    account: Account,
    intervals: IntervalSequence
  ): $ReadOnlyArray<Grain> {
    let allocationIndex = 0;
    return intervals.map((interval) => {
      let grain = G.ZERO;
      while (
        account.allocationHistory.length - 1 >= allocationIndex &&
        interval.startTimeMs <
          account.allocationHistory[allocationIndex].credTimestampMs &&
        account.allocationHistory[allocationIndex].credTimestampMs <=
          interval.endTimeMs
      ) {
        grain = G.add(
          grain,
          account.allocationHistory[allocationIndex].grainReceipt.amount
        );
        allocationIndex++;
      }
      return grain;
    });
  }

**Comment by bensodenkamp** [comment]
So as I get into the code I understand this problem space better. I don't think it is reasonable to try to really pin grain distributions to cred earned as a first step. If an org has used the balanced policy in the past, their grain distribution for a given time period will be for cred that is potentially before that time period, so if we try to go back and balance based on pinning the grain to the "right then" cred, we'll essentially be trying to unbalance what they already balanced.

I think a more reasonable approach is just to add a lookback parameter that works just like the balanced policy does now, but disregards imbalances before a certain point in time. More or less this would just be bumping up the start date for how the balanced policy sees the lifetime of the instance.

We could, in the future, go for a more 'grain is dated to associate with the cred it is awarded for,' but you would sort of have to start off that way, never having tried to balance across the full lifetime without pinning distributions to allocations.

Thoughts?
